required/latex-lab/ltprototype.dtx

```diff
@@ -0,0 +1,515 @@
+%   set by \cs{prototype_use_template:nnn} or
+%   \cs{prototype_declare_instance:nnnn}. If not default is given, the
+%   prevailing state when the template is used will apply.
+% \end{function}
+%
+% \section{Instances}
+% \label{sec:instances}
+%
+% After a template is defined it still needs to be put to use. The
+% parameters that it expects need to be defined before it can be used in
+% a document. Every time a template has parameters given to it, an
+% \emph{instance} is created, and this is the code that ends up in the
+% document to perform the typesetting of whatever pieces of information
+% are input into it.
+%
+% For example, a template might say \enquote{here is a section with or
+% without a number that might be centred or left aligned and print its
+% contents in a certain font of a certain size, with a bit of a gap
+% before and after it} whereas an instance declares \enquote{this is a
+% section with a number, which is centred and set in $12\,\text{pt}$
+% italic with a $10\,\text{pt}$ skip before and a
+% $12\,\text{pt}$ skip after it}. Therefore, an instance is just a
+% frozen version of a template with specific settings as chosen by the
+% designer.
+%
+% \begin{function}{\prototype_declare_instance:nnnn}
+%   \begin{syntax}
+%     \cs{prototype_declare_instance:nnnn}
+%     ~~\Arg{object type} \Arg{instance} \Arg{template} \Arg{parameters}
+%   \end{syntax}
+%   This function uses a \meta{template} for an \meta{object type}
+%   to create an \meta{instance}. The \meta{instance} will be set
+%   up using the \meta{parameters}, which will set some of the
+%   \meta{keys} in the \meta{template}.
+%
+%   As a practical example, consider an object type for document sections
+%   (which might include chapters, parts, sections, \emph{etc}.), which
+%   is called \texttt{sectioning}. One possible template for this
+%   object type might be called \texttt{basic}, and one instance of this
+%   template would be a numbered section. The instance declaration might
+%   read:
+%   \begin{verbatim}
+%     \prototype_declare_instance:nnnn { sectioning } { section-num } { basic }
+%       {
+%         numbered      = true ,
+%         justification = center ,
+%         font          = \normalsize\itshape ,
+%         before-skip   = 10pt ,
+%         after-skip    = 12pt ,
+%       }
+%   \end{verbatim}
+%   Of course, the key names here are entirely imaginary, but illustrate
+%   the general idea of fixing some settings.
+% \end{function}
+%
+% \section{Document interface}
+%
+% After the instances have been chosen, document commands must be
+% declared to use those instances in the document. \cs{UseInstance}
+% calls instances directly, and this command should be used internally
+% in document-level mark-up.
+%
+% \begin{function}{\prototype_use_instance:nn}
+%   \begin{syntax}
+%     \cs{prototype_use_instance:nn}
+%     ~~\Arg{object type} \Arg{instance} \meta{arguments}
+%   \end{syntax}
```

Comment:
needs to be discussed elsewhere but I think the instance (and template) use should accept one argument with one-time parameter settings, either always (in which case one would leave it empty if it shouldb't be used) or alternatively as two separate calls with different number of arguments